README.md
# Stork (pleroma-bot)

[![Build Status](https://travis-ci.com/robertoszek/pleroma-bot.svg?branch=develop)](https://app.travis-ci.com/github/robertoszek/pleroma-bot)
[![Version](https://img.shields.io/pypi/v/pleroma-bot.svg)](https://pypi.org/project/pleroma-bot/)
[![AUR version](https://img.shields.io/aur/version/python-pleroma-bot)](https://aur.archlinux.org/packages/python-pleroma-bot)
[![codecov](https://codecov.io/gh/robertoszek/pleroma-bot/branch/master/graph/badge.svg?token=0c4Gzv4HjC)](https://codecov.io/gh/robertoszek/pleroma-bot)
[![Python 3.6](https://img.shields.io/badge/python-3.6+-blue.svg)](https://www.python.org/downloads/release/python-360/)
[![License](https://img.shields.io/github/license/robertoszek/pleroma-bot)](https://github.com/robertoszek/pleroma-bot/blob/master/LICENSE.md)

![Stork](img/stork-smaller.svg)

Mirror your favorite Twitter accounts in the Fediverse, so you can follow their updates from the comfort of your favorite instance. Or migrate your own to the Fediverse using a Twitter [archive](https://twitter.com/settings/your_twitter_data).

[![Documentation](img/docs.png)](https://robertoszek.github.io/pleroma-bot)

You can find this project at: 

|                          GitHub                           |                          Gitlab                           |                                Gitea                                 |
|:---------------------------------------------------------:|:---------------------------------------------------------:|:--------------------------------------------------------------------:|
| [pleroma-bot](https://github.com/robertoszek/pleroma-bot) | [pleroma-bot](https://gitlab.com/robertoszek/pleroma-bot) | [pleroma-bot](https://gitea.robertoszek.xyz/robertoszek/pleroma-bot) |

Supports:

|                          Mastodon                           |                         Pleroma                         |                         Misskey                          |
|:-----------------------------------------------------------:|:-------------------------------------------------------:|:--------------------------------------------------------:|
| [![Mastodon](/img/mastodon.png)](https://joinmastodon.org/) | [![Pleroma](/img/pleroma.png)](https://pleroma.social/) | [![Misskey](/img/misskey.png)](https://misskey-hub.net/) |

## Introduction

After using the pretty cool [mastodon-bot](https://github.com/yogthos/mastodon-bot) for a while, I found it was lacking some features which were of use to me. 

For precisely those cases I've written this Python project that automates them, asking such info to [Twitter's API](https://developer.twitter.com/en/docs/twitter-api/v1) and updating the relevant fields on the [Pleroma API](https://api.pleroma.social/)/[Mastodon API](https://docs.joinmastodon.org/client/intro/)/[Misskey API](https://misskey-hub.net/en/docs/api) side.


## Features 

So basically, it does the following:
* Can parse a Twitter [archive](https://twitter.com/settings/your_twitter_data), moving all your tweets to the Fediverse
* Can use an RSS feed as the source of the tweets to post
* Retrieves latest **tweets** and posts them on the Fediverse account if their timestamp is newer than the last post.
  * Can filter out RTs or not
  * Can filter out replies or not
  * Supports Twitter threads
* Media retrieval and upload of multiple **attachments**. This includes:
  * Video
  * Images
  * Animated GIFs 
  * Polls
* Retrieves **profile info** from Twitter and updates it in on the Fediverse account. This includes:
  * *Display name*
  * *Profile picture*
  * *Banner image*
  * *Bio text*
* Adds some **metadata fields** to the Fediverse account, pointing to the original Twitter account or custom text.

## Installation
### Using pip
```
$ pip install pleroma-bot
```
### Using a package manager
Here's a list of the available packages.

| Package type   | Link                                                    | Maintainer                                    |
|:--------------:|:-------------------------------------------------------:|:---------------------------------------------:|
| AUR (Arch)     | https://aur.archlinux.org/packages/python-pleroma-bot  | [robertoszek](https://github.com/robertoszek) |

## Usage
```console
$ pleroma-bot [-c CONFIG] [-l LOG] [--noProfile] [--daemon] [--forceDate [FORCEDATE]] [-a ARCHIVE]
```

```console
Bot for mirroring one or multiple Twitter accounts in Pleroma/Mastodon.

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        path of config file (config.yml) to use and parse. If
                        not specified, it will try to find it in the current
                        working directory.
  -d, --daemon          run in daemon mode. By default it will re-run every
                        60min. You can control this with --pollrate
  -p POLLRATE, --pollrate POLLRATE
                        only applies to daemon mode. How often to run the
                        program in the background (in minutes). By default is
                        60min.
  -l LOG, --log LOG     path of log file (error.log) to create. If not
                        specified, it will try to store it at your config file
                        path
  -n, --noProfile       skips Fediverse profile update (no background image,
                        profile image, bio text, etc.)
  --forceDate [FORCEDATE]
                        forces the tweet retrieval to start from a specific
                        date. The twitter_username value (FORCEDATE) can be
                        supplied to only force it for that particular user in
                        the config
  -s, --skipChecks      skips first run checks
  -a ARCHIVE, --archive ARCHIVE
                        path of the Twitter archive file (zip) to use for
                        posting tweets.
  -t THREADS, --threads THREADS
                        number of threads to use when processing tweets
  -L LOCKFILE, --lockfile LOCKFILE
                        path of lock file (pleroma-bot.lock) to prevent
                        collisions with multiple bot instances. By default it
                        will be placed next to your config file.
  --verbose, -v
  --version             show program's version number and exit
```
### Before running

There are multiple options for using the bot.

You can either choose to use: 

- A Twitter archive
- An RSS feed
- Guest tokens
- [Twitter tokens](https://developer.twitter.com/en/docs/authentication/api-reference/token) with a Developer account 

You'll need to create a configuration file and obtain the [Fediverse tokens](https://tinysubversions.com/notes/mastodon-bot/) for your accounts no matter what you choose to use.

If you plan on retrieving tweets from an account which has their tweets **protected**, you'll also need the following:
* Consumer Key and Secret. You'll find them on your project app keys and tokens section at [Twitter's Developer Portal](https://developer.twitter.com/en/portal/dashboard)
* Access Token Key and Secret.  You'll also find them on your project app keys and tokens section at [Twitter's Developer Portal](https://developer.twitter.com/en/portal/dashboard). 
Alternatively, you can obtain the Access Token and Secret by running [this](https://github.com/joestump/python-oauth2/wiki/Twitter-Three-legged-OAuth-Python-3.0) locally, while being logged in with a Twitter account which follows or is the owner of the protected account

You'll may also need Elevated access in your Twitter's API project in order for the bot to function properly.

Refer to the docs [for more info about this](https://robertoszek.github.io/pleroma-bot/gettingstarted/beforerunning/#before-running).

### Configuration

Create a ```config.yml``` file in the same path where you are calling ```pleroma-bot``` (or use the `--config` argument to specify a different path). 

There's a config example in this repo called ```config.yml.sample``` that can help you when filling yours out.

For more information you can refer to the ["Configuration" page](https://robertoszek.github.io/pleroma-bot/gettingstarted/configuration/) on the docs.


CHANGELOG.md
## Unreleased
## Fixed
- Bug: Content warning text being empty for Misskey instances in some cases
- Bug: Unhandled error when unpinning if previous pinned post was deleted on Misskey. Thanks @ssmucny !
- Bug: Pinned post ID being saved on wrong path on Misskey instances
- Bug: Some signature links were not recognized as Twitter links by some third-party apps (like Twitter for Android). Thanks @wzyboy!

## Enhancements
- Progress bars will only be displayed on interactive TTYs/shells (removing unwanted noise from stderr on non-interactive shells like on cronjobs)

## [1.2.0] 02-01-2023
## Fixed
- Bug: Handle exception when media attachments are geoblocked (403 Unauthorized)
- Regression: Take into account new structure of Twitter's archive (tweet.js changed to tweets.js)
- Bug: Handle special media types that don't include link to media in some cases. Thanks @tewhalen!
- Bug: Align max_tweets minimum value (10) with the warning message and actual value. Thanks @nemobis!
- Bug: Multiple video attachments cause HTTP 422 on Mastodon when posting.
- Bug: Exception when post was manually pinned on the Fedi account and later one was pinned on the Twitter account. Thanks @selyod-ka!

## Enhancements
- Archive offline support, you no longer need a Twitter developer account or access to their API to process an archive.
- Mastodon limits, make URLs count as 23 characters (or the instance configured value) when determining if truncating is necessary
- Added progress bars when gathering, processing and posting tweets
- Pleroma and Mastodon rate limits, handle HTTP 429 more gracefully (wait until they reset and continue)

## Added
- RSS support, you can now use an RSS feed as a source of tweets to post. Take a look at the docs for more info.
- Thread support, now mirrored users should be able to reply to their own posts and create reply threads.
- Native retweet support, now users in your config should be able to reblog their own and each other's posts if they are on the same instance.
- `bot` config mapping, for setting the [bot flag](https://docs.joinmastodon.org/user/profile/#bot) on the target account.
- `--lockerfile` argument, for specifying the path of the locker file that prevents collisions between multiple concurrent bot executions.
- `guest` mapping, for enabling the use of Guest Tokens (so you don't need a Twitter Developer account). If no `twitter_token` is present in your config it will default to `true`
- `proxy_pool` mapping, list of proxies to use when being rate limited with Guest Tokens 
- `proxy` mapping, for enabling or disabling the use of proxies when using guest tokens
- Caching IDs of posts published and skip mirroring tweet if associated published post is found to avoid duplicates. You can change this with the `avoid_duplicates` mapping.
- `application_name` mapping, for specifying the Fediverse application name to use as a filter when getting the date of the latest published post by the bot. Thanks @reorx!
- Configuration wizard, which generates a minimal config if none is found
- `content_warnings`, for specifying keywords that if found will add a content warning to the Fediverse post
- `custom_replacements` for replacing any text (Twitter handles that differ from the Fedi handle, for example) Thanks @selyod-ka!
- `software` config mapping, for sidestepping the automatic detection of the type of software (mastodon, pleroma, misskey) running on the target instance and forcing a specific value. If your target is a fork of another project, for example, you would be able to force the bot to use whichever software (mastodon, pleroma, misskey) is the closest match for their API.

## [1.1.0] 06-06-2022
## Fixed
- Bug: Tweet media getting dropped if one of the attachments returned 404 Not Found
- Regression: `visibility`'s value not being honored when defined as a global mapping in the config
- Bug: Handle exception when the tweet is empty (no attachments, polls or body) due to dropping malformed attachments
- Bug: Handle exception when expanded URL is unreachable or returns any code other than 200

pleroma_bot/_processing.py
import os
import re
import json
import html
import shutil
import requests
import mimetypes
from datetime import datetime

from tqdm import tqdm

# Try to import libmagic
# if it fails just use mimetypes
try:
    import magic
except ImportError:
    magic = None

from . import logger
from .i18n import _


def process_tweets(self, tweets_to_post):
    """Transforms tweets for posting them to Pleroma
    Expands shortened URLs
    Downloads tweet related media and prepares them for upload

    :param tweets_to_post: Dict of tweet objects to be processed
    :type tweets_to_post: dict
    :returns: Tweets ready to be published
    :rtype: list
    """
    # TODO: Break into smaller functions

    # Remove RTs if include_rts is false
    if not self.include_rts:
        for tweet in tweets_to_post["data"][:]:
            try:
                for reference in tweet["referenced_tweets"]:
                    if reference["type"] == "retweeted":
                        tweets_to_post["data"].remove(tweet)
                        break
            except KeyError:
                pass
    # Remove quote tweets if include_quotes is false
    if not self.include_quotes:  # pragma: todo
        for tweet in tweets_to_post["data"][:]:
            try:
                for reference in tweet["referenced_tweets"]:
                    if reference["type"] == "quoted":
                        tweets_to_post["data"].remove(tweet)
                        break
            except KeyError:
                pass
    # Remove replies if include_replies is false
    if not self.include_replies:
        for tweet in tweets_to_post["data"][:]:
            try:
                for reference in tweet["referenced_tweets"]:
                    if reference["type"] == "replied_to":
                        tweets_to_post["data"].remove(tweet)
                        break
            except KeyError:
                pass

    if self.hashtags:
        for tweet in tweets_to_post["data"][:]:
            try:
                tweet_hashtags = tweet["entities"]["hashtags"]
                i = 0
                while i < len(tweet_hashtags):
                    if tweet_hashtags[i]["tag"] in self.hashtags:
                        break
                    i += 1
                else:
                    tweets_to_post["data"].remove(tweet)
            except KeyError:
                tweets_to_post["data"].remove(tweet)
                pass
    all_media = []
    if int(self.threads) == 1:
        desc = _("Processing tweets... ")
        pbar = tqdm(total=len(tweets_to_post["data"]), desc=desc, disable=None)

    for idx, tweet in enumerate(tweets_to_post["data"]):
        # logger.info(_("Processing: {}/{}").format(idx+1, total_tweets))
        posts = self.posts_ids[self.pleroma_base_url]
        if tweet["id"] not in posts:
            self.posts_ids[self.pleroma_base_url].update({tweet["id"]: ''})
        tweet["reply_id"] = None
        tweet["retweet_id"] = None
        # get reply ids
        if "referenced_tweets" in tweet.keys():
            for reference in tweet["referenced_tweets"]:
                if reference["type"] == "replied_to":
                    tweet["reply_id"] = reference["id"]
                if reference["type"] == "retweeted":
                    tweet["retweet_id"] = reference["id"]
        media = []
        logger.debug(tweet["id"])
        # Get full text from RT or quoted tweet
        if "referenced_tweets" in tweet.keys():  # pragma: no cover
            tweet["text"] = _get_rt_text(self, tweet)
        tweet["text"] = _expand_urls(self, tweet)
        tweet["text"] = html.unescape(tweet["text"])

        # Download media only if we plan to upload it later
        if self.media_upload and not self.archive:
            if self.guest:  # pragma: todo
                if "extended_entities" in tweet:
                    if "media" in tweet['extended_entities']:
                        for item in tweet['extended_entities']['media']:
                            media.append(item)
            try:
                media_keys = False
                attachments = "attachments" in tweet.keys()
                if attachments:
                    tweet_attachments = tweet["attachments"]
                    media_keys = "media_keys" in tweet_attachments.keys()
                if media_keys and attachments:
                    includes_media = tweets_to_post["includes"]["media"]
                    for item in tweet["attachments"]["media_keys"]:
                        for media_include in includes_media:
                            media_url = _get_media_url(
                                self, item, media_include, tweet
                            )
                            if media_url:
                                media.extend(media_url)
                # Get RT tweet media
                if "referenced_tweets" in tweet.keys():  # pragma: no cover
                    _get_rt_media_url(self, tweet, media)
            except KeyError:
                pass
            if len(media) > 0:
                # Create folder to store attachments related to the tweet ID
                tweet_path = os.path.join(self.tweets_temp_path, tweet["id"])
                os.makedirs(tweet_path, exist_ok=True)
                _download_media(self, media, tweet)
                all_media.extend(media)

        if not self.keep_media_links:
            tweet["text"] = _remove_media_links(self, tweet)
            tweet["text"] = _remove_status_links(self, tweet)
        if hasattr(self, "rich_text"):
            if self.rich_text:
                tweet["text"] = _replace_mentions(self, tweet)
        if self.nitter:
            tweet["text"] = _replace_url(
                self,
                tweet["text"],
                "https://twitter.com",
                self.nitter_base_url
            )
        if self.invidious:
            tweet["text"] = _replace_url(
                self,
                tweet["text"],
                "https://youtube.com",
                self.invidious_base_url
            )
        if hasattr(self, "custom_replacements"):
            if self.custom_replacements:
                tweet["text"] = _custom_replacements(
                    self,
                    tweet["text"],
                    self.custom_replacements
                )
        signature = ''
        if self.signature:
            if self.archive:  # pragma: todo
                t_user = self.twitter_ids[list(self.twitter_ids.keys())[0]]
            else:
                t_user = "_"
                if tweet["author_id"] in self.twitter_ids:
                    t_user = self.twitter_ids[tweet["author_id"]]
            twitter_url_user = f"{self.twitter_url_home}/{t_user}"
            signature = f"\n\n 🐦🔗: {twitter_url_user}/status/{tweet['id']}"
            if self.instance == "mastodon":
                len_text = self._mastodon_len(tweet["text"])
                len_signature = self._mastodon_len(signature)
            else:
                len_text = len(tweet["text"])
                len_signature = len(signature)
            total_length = len_text + len_signature
            if total_length > self.max_post_length:  # pragma
                body_max_length = self.max_post_length - len_signature - 1
                tweet["text"] = f"{tweet['text'][:body_max_length]}…"
            tweet["text"] = f"{tweet['text']}{signature}"
        if self.original_date:
            tweet_date = tweet["created_at"]
            date = datetime.strftime(
                datetime.strptime(tweet_date, "%Y-%m-%dT%H:%M:%S.000Z"),
                self.original_date_format,
            )
            orig_date = f"\n\n[{date}]"
            if self.instance == "mastodon":
                len_text = self._mastodon_len(tweet["text"])
            else:
                len_text = len(tweet["text"])
            total_length = len_text + len(orig_date)
            if total_length > self.max_post_length:  # pragma
                if self.signature:
                    tweet["text"] = tweet["text"].replace(signature, '')
                l_date = len(orig_date)
                if self.instance == "mastodon":
                    l_sig = self._mastodon_len(signature)
                else:
                    l_sig = len(signature)
                body_max_length = self.max_post_length - l_date - l_sig - 1
                tweet["text"] = f"{tweet['text'][:body_max_length]}…"
            else:
                signature = ''
            tweet["text"] = f"{tweet['text']}{signature}{orig_date}"
        # Process poll if exists and no media is used
        tweet["polls"] = None
        if not self.guest:
            tweet["polls"] = _process_polls(self, tweet, media)

        # Truncate text if needed
        if self.instance == "mastodon":
            total_tweet_length = self._mastodon_len(tweet["text"])
        else:
            total_tweet_length = len(tweet["text"])
        if total_tweet_length > self.max_post_length:  # pragma
            logger.info(
                _(
                    "Post text longer than allowed ({}), truncating..."
                ).format(self.max_post_length)
            )
            tweet["text"] = f"{tweet['text'][:self.max_post_length]}"
        tweet["cw"] = None
        if self.content_warnings:
            tweet["cw"] = _check_cw(tweet["text"], self.content_warnings)
        if int(self.threads) == 1:
            pbar.update(1)
    tweets_to_post["media_processed"] = all_media
    return tweets_to_post


def _check_cw(data, cw_list):
    cw_found = []
    for cw_topic in cw_list:
        alnum = [k for k in cw_list[cw_topic] if k.isalnum()]
        not_alnum = [k for k in cw_list[cw_topic] if not k.isalnum()]
        topic_found = False
        if not_alnum:
            matches = re.findall("|".join(not_alnum), data, re.IGNORECASE)
            if matches:
                topic_found = True
        if any(k.lower() in data.lower().split() for k in alnum):
            topic_found = True
        if topic_found:
            cw_found.append(cw_topic.lower())
    cw_text = ", ".join(cw_found).capitalize()
    return cw_text


def _custom_replacements(self, data, custom_replacements):
    pairs = custom_replacements.items()
    # normalize dict keys to lower-case for case-insensitive matching
    custom_replacements = {k.lower(): v for k, v in pairs}
    # alternation regex on all keys in custom_replacements
    keys_regex = "|".join(custom_replacements.keys())
    matches = re.findall(keys_regex, data, re.IGNORECASE)
    for match in matches:
        data = re.sub(
            match, lambda m: custom_replacements[m.group().lower()], data
        )
    return data


def _get_rt_text(self, tweet):  # pragma: no cover
    text = tweet["text"]

    for reference in tweet["referenced_tweets"]:
        retweeted = reference["type"] == "retweeted"
        quoted = reference["type"] == "quoted"
        if retweeted or quoted:
            tweet_ref_id = reference["id"]
            tweet_ref = self._get_tweets("v2", tweet_ref_id)
            if not tweet_ref:  # pragma: todo
                continue
            match = re.search(r"RT.*?\:", tweet["text"])
            prefix = match.group() if match else ""
            if retweeted:
                text = f"{prefix} {tweet_ref['data']['text']}"
            elif quoted:
                author = tweet_ref["data"]["author_id"]
                users = tweet_ref["includes"]["users"]
                username = [u["username"] for u in users if author == u["id"]]
                prefix = f"RT @{username[0]}:"
                text = f"{tweet['text']}\n{prefix} {tweet_ref['data']['text']}"
        else:
            break
    return text


def _get_rt_media_url(self, tweet, media):  # pragma: no cover
    tweet_rt = {"data": tweet}
    tw_data = tweet_rt["data"]
    i = 0
    max_at = self.max_attachments
    while "referenced_tweets" in tw_data.keys() and len(media) < max_at:
        for reference in tw_data["referenced_tweets"]:
            retweeted = reference["type"] == "retweeted"
            quoted = reference["type"] == "quoted"
            if retweeted or quoted:
                tweet_id = reference["id"]
                tweet_rt = self._get_tweets("v2", tweet_id)
                if not tweet_rt:
                    continue
                tw_data = tweet_rt["data"]
                att = "attachments" in tw_data.keys()
                if att:
                    attachments = tw_data["attachments"]
                    in_md = tweet_rt["includes"]["media"]
                    md_keys = attachments["media_keys"]
                    for item in md_keys:
                        for media_include in in_md:
                            media_url = _get_media_url(
                                self,
                                item,
                                media_include,
                                tw_data
                            )
                            if media_url:
                                new = [
                                    i for i in media_url if i["url"] not in [
                                        j["url"] for j in media
                                    ]
                                ]
                                media.extend(new)
                if self.guest:  # pragma: todo
                    if "extended_entities" in tw_data:
                        if "media" in tw_data['extended_entities']:
                            for item in tw_data['extended_entities']['media']:
                                if item not in media:
                                    media.append(item)
            else:
                break
        i += 1
        if i > 3:
            logger.debug(
                _("Giving up, reference is too deep")
            )
            break


def _process_polls(self, tweet, media):
    tweet["polls"] = None
    try:
        if tweet["attachments"]["poll_ids"] and not media:

            # tweet_poll = tweet['includes']['polls']
            poll_url = f"{self.twitter_base_url_v2}/tweets"

            params = {
                "ids": tweet["id"],
                "expansions": "attachments.poll_ids",
                "poll.fields": "duration_minutes," "options",
            }

            response = self.twitter_api_request(
                'GET',
                poll_url,
                headers=self.header_twitter,
                params=params,
                auth=self.auth
            )
            if not response.ok:
                response.raise_for_status()
            response_content = json.loads(response.content)
            tweet_poll = response_content["includes"]["polls"][0]

            pleroma_poll = {
                "options": [
                    option["label"] for option in tweet_poll["options"]
                ],
                "expires_in": tweet_poll["duration_minutes"] * 60,
            }

            # Add poll to tweet
            tweet["polls"] = pleroma_poll

        else:
            tweet["polls"] = None
    except KeyError:
        tweet["polls"] = None
        pass

    return tweet["polls"]


def _download_media(self, media, tweet):
    for idx, item in enumerate(media):
        if item["type"] != "video" and item["type"] != "animated_gif":
            if "media_url" in item:  # pragma: todo
                media_url = item['media_url']
                if len(media_url) == 0 and "media_url_https" in item:
                    media_url = item['media_url_https']
            else:
                media_url = item["url"]
        else:
            media_url = _get_best_bitrate_video(self, item)

        if media_url:
            if "media_key" not in item:  # pragma: todo
                item["media_key"] = str(item["id"])
            key = item["media_key"]
            response = requests.get(media_url, stream=True)
            try:
                if not response.ok:
                    response.raise_for_status()
            except requests.exceptions.HTTPError:
                if response.status_code == 404:
                    att_not_found = _(
                        "Exception occurred"
                        "\nMedia not found (404)"
                        "\n{tweet} - {media_url}"
                        "\nIgnoring attachment and continuing..."
                    ).format(tweet=tweet["id"], media_url=media_url)
                    logger.warning(att_not_found)
                    continue
                elif response.status_code == 403:
                    geoblocked = _(
                        "Media possibly geoblocked? (403) Skipping... "
                        "{tweet} - {media_url} "
                    ).format(tweet=tweet["id"], media_url=media_url)
                    logger.warning(geoblocked)
                    continue
                else:
                    response.raise_for_status()
            response.raw.decode_content = True
            filename = str(idx) + "-" + str(key) + mimetypes.guess_extension(
                response.headers["Content-Type"]
            )
            file_path = os.path.join(
                self.tweets_temp_path, tweet["id"], filename
            )
            with open(file_path, "wb") as outfile:
                shutil.copyfileobj(response.raw, outfile)
            # Remove attachment if exceeds the limit
            if hasattr(self, "file_max_size"):
                file_size_bytes = os.stat(file_path).st_size
                max_file_size_bytes = parse_size(self.file_max_size)
                if file_size_bytes > max_file_size_bytes:
                    logger.error(
                        _(
                            "Attachment exceeded config file size limit ({})"
                        ).format(self.file_max_size)
                    )
                    logger.error(
                        _("File size: {}MB").format(
                            round(file_size_bytes / 2 ** 20, 2)
                        )
                    )
                    logger.error(_("Ignoring attachment and continuing..."))
                    os.remove(file_path)


def parse_size(size):
    units = {
        "B": 1,
        "KB": 2 ** 10,
        "MB": 2 ** 20,
        "GB": 2 ** 30,
        "TB": 2 ** 40,
    }
    size = size.upper()
    if not re.match(r" ", size):
        size = re.sub(r"([KMGT]?B)", r" \1", size)
    number, unit = [string.strip() for string in size.split()]
    return int(float(number) * units[unit])


def _replace_url(self, data, url, new_url):
    matches = re.findall(url, data)
    for match in matches:
        data = re.sub(match, new_url, data)
    return data


def _remove_status_links(self, tweet):
    regex = r"\bhttps?:\/\/twitter.com\/+[^\/:]+\/.*?status\/\d*\b"
    tweet["text"] = re.sub(regex, '', tweet["text"], 0, re.MULTILINE)
    return tweet["text"]


def _remove_media_links(self, tweet):
    regex = r"\bhttps?:\/\/twitter.com\/+[^\/:]+\/.*?(photo|video)\/\d*\b"
    tweet["text"] = re.sub(regex, '', tweet["text"], 0, re.MULTILINE)
    return tweet["text"]


def _replace_mentions(self, tweet):
    matches = re.findall(r"\B\@\w+", tweet["text"])
    for match in matches:
        # TODO: Use nitter if asked (self.nitter)
        mention_link = f"[{match}](https://twitter.com/{match[1:]})"
        tweet["text"] = re.sub(match, mention_link, tweet["text"])
    return tweet["text"]


def _expand_urls(self, tweet):
    # TODO: transform twitter links to nitter links, if self.nitter
    #  'true' in resolved shortened urls

    # URI regex
    matching_pattern = (
        r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]"
        r"{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*"
        r"\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{}"
        r';:\'".,<>?«»“”‘’]))'
    )
    matches = re.finditer(matching_pattern, tweet["text"])
    urls = {}
    # Prepare a session to recycle connections across requests
    session = requests.Session()
    # Replace shortened links
    for matchNum, match in enumerate(matches, start=1):
        group = match.group()
        try:
            if len(tweet["entities"]["urls"]) > 0:
                entities_urls = tweet["entities"]["urls"]
                urls = {
                    u["url"]: u["expanded_url"] for u in entities_urls
                }
            if group in urls.keys():
                tweet["text"] = re.sub(group, urls[group], tweet["text"])
            else:
                raise KeyError
        except (KeyError, TypeError):
            # don't be brave trying to unwound an URL when it gets cut off
            if (
                    not group.__contains__("…") and not
                    group.startswith(self.nitter_base_url)
            ):
                if not group.startswith(("http://", "https://")):
                    group = f"http://{group}"
                try:
                    response = session.head(
                        group,
                        allow_redirects=True,
                        timeout=(7, 10)
                    )
                    if not response.ok:
                        logger.debug(
                            _(
                                "Couldn't expand the url {}: {}"
                            ).format(group, response.status_code)
                        )
                    if response:
                        expanded_url = response.url
                        tweet["text"] = re.sub(
                            group, expanded_url, tweet["text"]
                        )
                except Exception as ex:  # pragma
                    logger.debug(
                        _(
                            "Couldn't expand the url: {}"
                        ).format(group)
                    )
                    if isinstance(ex, requests.exceptions.RequestException):
                        expanded_url = ex.request.url
                        tweet["text"] = re.sub(
                            group, expanded_url, tweet["text"]
                        )
                        pass
                    elif isinstance(ex, UnicodeError):
                        expanded_url = ex.object.decode('latin1')
                        tweet["text"] = re.sub(
                            group, expanded_url, tweet["text"]
                        )
                        pass
    return tweet["text"]


def _get_media_url(self, item, media_include, tweet):
    # TODO: Verify if video download is available on v2 and migrate to it
    media_urls = []
    if item == media_include["media_key"]:
        # Video download not implemented in v2 yet
        # fallback to v1.1
        if (
            media_include["type"] == "video"
            or media_include["type"] == "animated_gif"
        ):
            tweet_video = self._get_tweets("v1.1", tweet["id"])
            xmd = tweet_video["extended_entities"]["media"]
            for extended_media in xmd:
                extended_media["media_key"] = item
                media_urls.append(extended_media)
            return media_urls
        else:
            media_urls.append(media_include)
            return media_urls


def _get_best_bitrate_video(self, item):
    bitrate = 0
    url = ""
    for variant in item["video_info"]["variants"]:
        try:
            if "bitrate" in variant:
                if int(variant["bitrate"]) >= bitrate:
                    bitrate = int(variant["bitrate"])
                    url = variant["url"]
            else:
                continue
        except KeyError:  # pragma: no cover
            pass
    return url
